json_rpc/servers/httpserver.nim

```diff
@@ -1,233 +1,64 @@
-          # Timeout
-          debug "Timeout expired while receiving request body",
-                address = transp.remoteAddress()
-          discard await transp.sendAnswer(header.version, Http413)
-          await transp.closeWait()
-          break
+    httpServers: seq[HttpServerRef]
+
+proc addHttpServer*(rpcServer: RpcHttpServer, address: TransportAddress) =
+  proc processClientRpc(rpcServer: RpcHttpServer): HttpProcessCallback {.closure.} =
+    return proc (req: RequestFence): Future[HttpResponseRef] {.async.} =
+      if req.isOk():
+        let request = req.get()
+        let body = await request.getBody()
+
+        let future = rpcServer.route(cast[string](body))
+        yield future
+        if future.failed:
+          debug "Internal error while processing JSON-RPC call"
+          return await request.respond(Http503, "Internal error while processing JSON-RPC call")
         else:
-          blenfut.read()
-      except TransportIncompleteError:
-        # remote peer disconnected
-        debug "Remote peer disconnected", address = transp.remoteAddress()
-        await transp.closeWait()
-        break
-      except TransportOsError as exc:
-        debug "Problems with networking", address = transp.remoteAddress(),
-              error = exc.msg
-        await transp.closeWait()
-        break
-
-      let future = rpc.route(cast[string](buffer))
-      yield future
-      if future.failed:
-        # rpc.route exception
-        debug "Internal error, while processing RPC call",
-              address = transp.remoteAddress()
-        let res = await transp.sendAnswer(header.version, Http503)
-        if not res:
-          await transp.closeWait()
-          break
+          var data = future.read()
+          let res = await request.respond(Http200, data)
+          trace "JSON-RPC result has been sent"
+          return res
       else:
-        var data = future.read()
-        let res = await transp.sendAnswer(header.version, Http200, data)
-        trace "RPC result has been sent", address = $transp.remoteAddress()
-        if not res:
-          await transp.closeWait()
-          break
-    elif vres == ErrorFailure:
-      debug "Remote peer disconnected", address = transp.remoteAddress()
-      await transp.closeWait()
-      break
-
-    if header.version in {HttpVersion09, HttpVersion10}:
-      debug "Disconnecting client", address = transp.remoteAddress()
-      await transp.closeWait()
-      break
-    else:
-      if connection == "close":
-        debug "Disconnecting client", address = transp.remoteAddress()
-        await transp.closeWait()
-        break
-
-  debug "Finished connection", address = $transp.remoteAddress()
+        return dumbResponse()
 
-# Utility functions for setting up servers using stream transport addresses
-
-proc addStreamServer*(server: RpcHttpServer, address: TransportAddress) =
   try:
     info "Starting JSON-RPC HTTP server", url = "http://" & $address
```

Comment:
What about secure (https) servers? This should be now possible with Chronos.

Reply:
ok I'm working on a HTTPS Server implementation.
should I do the HTTPS Client too?

Reply:
e9be98e9368b30f10cac4c4eb861a5dd40faabc6